### PR TITLE
OTP authentication draft

### PR DESCRIPTION
**Pull request by Foxboron**
This introduces a WIP OTP authentication login mechanism for the
tracker. This introduces a hack, and proves a possible implementation
error.

The views themselves can not change the `is_authorized` attribute. Only
the `load_user` function in `user.py` does this as our only check is on
`current_user`. This function is only creating `current_user` after the
initial `login_user(user)` call done after successfully verification.

Any calls to `login_user(user)` will authenticate the session and no
partial session can be created. To get around this we check if OTP
authentication is enabled, then create a session variable `two_factor`
which is checked inside the view and form to prompt further
authentication.

Signed-off-by: Morten Linderud <morten@linderud.pw>

### CODE REVIEW

File: tracker/model/user.py [hunk 1]
```diff
@@ -22,10 +23,13 @@ class User(db.Model):
     token = db.Column(db.String(TOKEN_LENGTH), index=True, unique=True, nullable=True)
     role = db.Column(UserRole.as_type(), nullable=False, default=UserRole.reporter)
     active = db.Column(db.Boolean(), nullable=False, default=True)
+    otp_token = db.Column(db.String(OTP_TOKEN_LENGTH), index=True, nullable=True)
 
     is_authenticated = False
     is_anonymous = False
 
+    first_time_login_otp = False
```
Comment (Foxboron) [inline tracker/model/user.py]: This isn't really used :x

File: tracker/view/login.py [hunk 2]
```diff
@@ -29,9 +35,36 @@ def login():
...
+        user.otp_token = pyotp.random_base32()
+        user.first_time_login_otp = True
+    form = OTPForm()
+    if not form.validate_on_submit():
+        status_code = Unauthorized.code if form.is_submitted() else 200
+        return render_template('two_factor.html',
+                               title='Two Factor Login',
+                               User=User,
+                               form=form,
+                               display_qr=current_user.first_time_login_otp,
```
Comment (Foxboron) [inline tracker/view/login.py]: This isn't going to work anymore as `current_user` isn't a source for information unless we call `login_user(user)`.

File: requirements.txt [hunk 3]
```diff
@@ -7,3 +7,5 @@ flask-talisman
 requests
 scrypt
 https://git.archlinux.org/pyalpm.git/snapshot/pyalpm-0.8.4.tar.gz
+Flask-QRcode
+pyotp
```
Comment (jelly) [inline requirements.txt]: Going to need some packaging in [community] :)

**Comment by jelly** [comment]
I think you need to generate a migration for the User model change

File: tracker/templates/two_factor.html [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{%- extends "base.html" -%}
+{% block content %}
+			<h1>Two Factor</h1>
+            {% if display_qr %}
+                <img src="{{ qrcode(otp_string) }}">
+            {%endif%}
```
Comment (jelly) [inline tracker/templates/two_factor.html]: Remove trailing spaces

File: tracker/model/user.py (hunk 1 above)
Comment (jelly) [inline tracker/model/user.py]: Should be true, since users don't have an token the first time they are created.

File: tracker/__init__.py [hunk 5]
```diff
@@ -66,7 +67,8 @@ def db_get_or_create(self, model, defaults=None, **kwargs):
     'default-src': '\'self\'',
     'style-src': '\'self\'',
     'font-src': '\'self\'',
-    'form-action': '\'self\''
+    'form-action': '\'self\'',
+    'img-src': 'data:'
```
Comment (jelly) [inline tracker/__init__.py]: This now disallows our Arch logo, so needs fixing.

Comment (jelly) [reply tracker/__init__.py]: Should be: ```'img-src': '\'self\' data:'```

File: tracker/form/login.py [hunk 6]
```diff
@@ -1,3 +1,5 @@
+import pyotp
```
Comment (anthraxx) [inline tracker/form/login.py]: please always use `from bla import foo` imports and make sure isort is in line via `make isort`

File: tracker/form/login.py [hunk 7]
```diff
@@ -45,3 +52,28 @@ def fail():
             return False
         self.user = user
         return True
+
+
+class OTPForm(BaseForm):
+    otp_code = StringField(u'OTP Code', validators=[DataRequired(), Length(max=6)])
```
Comment (anthraxx) [inline tracker/form/login.py]: min length would be a good idea as well

File: tracker/model/user.py [hunk 8]
```diff
@@ -22,10 +23,13 @@ class User(db.Model):
     token = db.Column(db.String(TOKEN_LENGTH), index=True, unique=True, nullable=True)
     role = db.Column(UserRole.as_type(), nullable=False, default=UserRole.reporter)
     active = db.Column(db.Boolean(), nullable=False, default=True)
+    otp_token = db.Column(db.String(OTP_TOKEN_LENGTH), index=True, nullable=True)
```
Comment (anthraxx) [inline tracker/model/user.py]: we don't need an index on this field

File: tracker/templates/two_factor.html (hunk 4 above)
Comment (anthraxx) [reply tracker/templates/two_factor.html]: please also use tabs in html and css files

File: tracker/view/login.py [hunk 9]
```diff
@@ -1,15 +1,21 @@
+import pyotp
```
Comment (anthraxx) [inline tracker/view/login.py]: please always use `from bla import foo` imports and make sure isort is in line via `make isort`

File: tracker/view/login.py [hunk 10]
```diff
@@ -29,9 +35,36 @@ def login():
...
+    if TRACKER_ENABLE_TWO_FACTOR:
+        session['two_factor'] = form.user.name
+        return redirect(url_for('tracker.two_factor'))
     user = user_assign_new_token(form.user)
-    user.is_authenticated = True
+    login_user(user)
+    return redirect(url_for('tracker.index'))
+
+
+@tracker.route('/twofactor', methods=['GET', 'POST'])
```
Comment (anthraxx) [inline tracker/view/login.py]: I don't like this being a separate endpoint, all this logic should be in the login endpoint as there is no reason to have a separate accessible URL for this, its the second step of the same process: login

File: tracker/view/login.py [hunk 11]
```diff
@@ -29,9 +35,36 @@ def login():
...
+    form = OTPForm()
+    if not form.validate_on_submit():
+        status_code = Unauthorized.code if form.is_submitted() else 200
+        return render_template('two_factor.html',
+                               title='Two Factor Login',
+                               User=User,
+                               form=form,
+                               display_qr=user.first_time_login_otp,
+                               otp_string=TRACKER_OTP_FORMAT.format(user.name, user.otp_token),
+                               otp_code_length={'max': 6}), status_code
```
Comment (anthraxx) [inline tracker/view/login.py]: we repeat the magic value of 6 another (once in the form). Please move this to the config even if it should always be 6

File: tracker/form/login.py [hunk 12]
```diff
@@ -45,3 +52,28 @@ def fail():
...
+            return False
+
+        def fail():
+            self.otp_code.errors.append(ERROR_INVALID_OTP)
+            return False
+
+        if not session.get('two_factor'):
+            return False
+
+        user = User.query.filter(User.name == session['two_factor']).first()
```
Comment (anthraxx) [inline tracker/form/login.py]: Please don't do any additional queries on the User object other then using `current_user` and call `db.session.commit()` if some database backed field was changed.

File: config.py [hunk 13]
```diff
@@ -34,6 +34,8 @@ def set_debug_flag(debug):
 TRACKER_PASSWORD_LENGTH_MIN = config_tracker.getint('password_length_min')
 TRACKER_PASSWORD_LENGTH_MAX = config_tracker.getint('password_length_max')
 TRACKER_SUMMARY_LENGTH_MAX = config_tracker.getint('summary_length_max')
+TRACKER_OTP_FORMAT = config_tracker['otp_format']
+TRACKER_ENABLE_TWO_FACTOR = config_tracker.getboolean('enable_two_factor')
```
Comment (anthraxx) [inline config.py]: both should be named `enforce_two_factor` as 2FA should always be possible its just a matter of configuration if it is enforced on everyone or not.

File: tracker/model/user.py (hunk 1 above)
Comment (anthraxx) [reply tracker/model/user.py]: I don't think we need this property

**Comment by jelly** [comment]
I guess we will defer this PR, as we want to get Keycloak and the arch-security-tracker working together soon. And then enforce 2FA on Keycloak.

**Comment by svenstaro** [comment]
Closing this in favor of keycloak integration.

**Pull request closed without merging**